addons/account/models/account_invoiceable_order.py

```diff
@@ -0,0 +1,226 @@
+
+    def _qty_field_name(self):
+        """
+        Returns a string with the name of the quantity field on the implementing subclass.
+        Used in creating the down payment lines values.
+        """
+        raise NotImplementedError  # To override
+
+    def _sign(self):
+        """
+        Returns the sign (1 or -1) to multiply with when calculating the amount to invoice and creating the invoice lines.
+        """
+        raise NotImplementedError  # To override
+
+    def _is_locked(self):
+        """
+        Returns whether the order is locked.
+        Used to decide which down payment lines to update the status of when posting a linked down payment invoice.
+        """
+        raise NotImplementedError  # To override
+
+    def _filter_subrecords(self):
+        """
+        Returns a filter function that only returns True for invoice lines linked to this order.
+        Used to compute the amount to invoice.
+        """
+        raise NotImplementedError  # To override
+
+    def _subrecord_type(self):
+        """
+        Returns an empty recordset of the subrecord (sale.order.line or purchase.order.line).
+        Used to create the order lines of the correct type.
+        """
+        raise NotImplementedError  # To override
+
+    def _prepare_invoice(self):
+        """
+        Prepare the dict of values to create the new invoice for this order.
+        Used when generating the invoice from this invoiceable order.
+        """
+        raise NotImplementedError  # To override
+
+    def _create_invoices(self, grouped=False, final=False, date=None):
+        """
+        Create the invoices for this invoiceable order.
+        """
+        raise NotImplementedError  # To override
+
+    def action_view_invoice(self, invoices=False):
+        """
+        Returns the action to view the provided invoices.
+        Used to display the created invoices after creation.
+        """
+        raise NotImplementedError  # To override
+
+    def _get_lang(self):
+        self.ensure_one()
+
+        if self.partner_id.lang and not self.partner_id.is_public:
+            return self.partner_id.lang
+
+        return self.env.lang
+
+    @api.depends('invoice_ids.state', 'currency_id', 'amount_total')
+    def _compute_amount_to_invoice(self):
+        for order in self:
+            # If the invoice status is 'Fully Invoiced' force the amount to invoice to equal zero and return early.
+            if order.invoice_status == 'invoiced':
+                order.amount_to_invoice = 0.0
+                continue
+
+            order.amount_to_invoice = order.amount_total
+            for invoice in order.invoice_ids.filtered(lambda x: x.state == 'posted'):
+                prices = sum(invoice.line_ids.filtered(order._filter_subrecords()).mapped('price_total'))
+                invoice_amount_currency = invoice.currency_id._convert(
+                    prices * invoice.direction_sign * order._sign(),
```

Comment:
I would be hesitant to change this as there is indeed the complexity with credit notes, taxes and rounding to take into account. There was already a bugfix because it was not working properly in the credit note case...

Reply:
Alright, let's forget that for now. Just keep it in mind for the future, maybe :thinking: